streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -111,12 +112,14 @@ public StreamTask(final TaskId id,
                       final StateDirectory stateDirectory,
                       final ThreadCache cache,
                       final Time time,
-                      final Producer<byte[], byte[]> producer) {
+                      final Producer<byte[], byte[]> producer,
+                      final Sensor skippedRecordsSensor) {
```

Comment:
Again, if we could pass around the `threadMetrics` here, it will make the code more readable: we can make it very clear at which places we record some task metrics like `taskMetrics.sensorA.record()` and where do we record thread-level metrics like `threadMetrics.skippedRecordsSensor.record()`.

But I think it is better to be left as a follow-up PR as this one is already pretty big.

Reply:
Yeah, I keep getting wrapped around the axle thinking about stuff like this. Hopefully, I'll be able to deliver a reasonable implementation for this PR, and I'll continue to mull about a way to pass the right metric context around the code base.